packages/utils/src/lib/reports/generate-md-reports-diff.ts

```diff
@@ -0,0 +1,297 @@
+import { AuditDiff, Commit, ReportsDiff } from '@code-pushup/models';
+import { pluralize, pluralizeToken } from '../formatting';
+import { objectToEntries } from '../transform';
+import { Alignment, details, h1, h2, paragraphs, style, tableMd } from './md';
+import { DiffOutcome } from './types';
+import {
+  colorByScoreDiff,
+  formatDiffNumber,
+  formatScoreWithColor,
+  getDiffMarker,
+  getSquaredScoreMarker,
+} from './utils';
+
+// to prevent exceeding Markdown comment character limit
+const MAX_ROWS = 100;
+
+export function generateMdReportsDiff(diff: ReportsDiff): string {
+  return paragraphs(
+    formatDiffHeaderSection(diff),
+    formatDiffCategoriesSection(diff),
+    formatDiffGroupsSection(diff),
+    formatDiffAuditsSection(diff),
```

Comment:
These could have their own unit tests.

Reply:
These functions are already covered extensively by the integration tests, I don't see how unit testing them again separately would improve the test coverage. A lot of data is needed in the report diff, so crafting the test cases is very time-consuming. And it's more beneficial to test it holistically because the Markdown output is inherently visual.

Is there some test case which you would like to see covered, which isn't already covered by the integration tests?

Reply:
Well, the general idea is to have a test written for as small a scope as possible. The test data could then be much simpler as well (only categories or only audits). So the question is more the other way around - what can we already cover with unit tests that we wouldn't need to create integration test cases for (only keep a happy path).

But I suppose this would require more effort so it is okay for now.